server/webapp/WEB-INF/rails.new/webpack/models/artifact_stores/artifact_stores.js

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2018 ThoughtWorks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const Stream               = require('mithril/stream');
+const s                    = require('string-plus');
+const Mixins               = require('models/mixins/model_mixins');
+const PluginConfigurations = require('models/shared/plugin_configurations');
+const SparkRoutes          = require('helpers/spark_routes');
+const Validatable          = require('models/mixins/validatable_mixin');
+const CrudMixins           = require('models/mixins/crud_mixins');
+
+const ArtifactStores = function (data) {
+  Mixins.HasMany.call(this, {
+    factory:    ArtifactStores.ArtifactStore.create,
+    as:         'ArtifactStore',
+    collection: data,
+    uniqueOn:   'id'
+  });
+};
+
+ArtifactStores.API_VERSION = 'v1';
+
+CrudMixins.Index({
+  type:     ArtifactStores,
+  indexUrl: SparkRoutes.artifactStoresPath(),
+  version:  ArtifactStores.API_VERSION,
+  dataPath: '_embedded.artifact_stores'
+});
+
+
+ArtifactStores.ArtifactStore = function (data) {
+  this.id         = Stream(s.defaultToIfBlank(data.id, ''));
+  this.pluginId   = Stream(s.defaultToIfBlank(data.pluginId, ''));
+  this.properties = s.collectionToJSON(Stream(s.defaultToIfBlank(data.properties, new PluginConfigurations())));
+  this.parent     = Mixins.GetterSetter();
+  this.etag       = Mixins.GetterSetter();
+  Mixins.HasUUID.call(this);
+
+  Validatable.call(this, data);
+
+  this.validatePresenceOf('id');
+  this.validatePresenceOf('pluginId');
+  this.validateFormatOf('id', {
+    format:  /^[-a-zA-Z0-9_][-a-zA-Z0-9_.]*$/,
+    message: 'Invalid id. This must be alphanumeric and can contain underscores and periods (however, it cannot start ' +
```

Comment:
Can we add this to `constants.js`? We seem to be using this message in a lot of places.

Reply:
I've moved it to the [validatable mixin](https://github.com/gocd/gocd/pull/4628/commits/ce32a003359cb6248df00dce6b56f7856b0afe5e) instead. Seemed like a better place for it.